### PR TITLE
block cron when in single user mode

### LINKED ISSUE #15498: Cron.php runs in single-user mode
### Steps to reproduce
Following similar steps as https://github.com/owncloud/core/issues/14843
1. Set single-user mode in the server
2. `update oc_jobs set last_run=0`
3. Run `php cron.php`
4. Check "oc_jobs" table

### Expected behaviour
Cron doesn't run, timestamps in the "oc_jobs" table don't change

### Actual behaviour
Cron runs

### Server configuration
**Operating system**: ubuntu 14.04

**Web server:** apache 2.4.7

**Database:** mysql

**PHP version:** 5.5.4

**ownCloud version:** 8.0.3RC1

**Updated from an older ownCloud or fresh install:** fresh install

### PR DESCRIPTION
**Pull request by icewind1991**
Fixes #15498

cc @PVince81 @jvillafanez

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
Please double check all places where we check for maintenance mode and apply this check as well - THX

**Comment by nickvergessen** [comment]
> Please double check all places where we check for maintenance mode and apply this check as well - THX

https://github.com/owncloud/core/issues/15505
Relevant PR from yesterday https://github.com/owncloud/core/pull/15465/files

**Comment by karlitschek** [comment]
:+1:

**Comment by nickvergessen** [comment]
and https://github.com/owncloud/core/issues/15495

**Comment by icewind1991** [comment]
Now also blocks webdav and ocs

**Comment by PVince81** [comment]
This will need to be added to caldav and carddav too. Ah, if only we had only a single entry point for everything.

Testing single user mode:
- [ ] cron.php
- [ ] OCS call
- [ ] remote.php/webdav
- [ ] public.php
- [ ] publicwebdav

**Comment by DeepDiver1975** [comment]
> FAIL I think we should block public links too ?

we don't do that for maintenance mode afaik

**Comment by PVince81** [comment]
Yes we do: Public links are not accessible when in maintenance mode.

**Comment by DeepDiver1975** [comment]
> Yes we do: Public links are not accessible when in maintenance mode.

really - wow ...

**Comment by DeepDiver1975** [comment]
> really - wow ...

this should already work ... https://github.com/owncloud/core/blob/master/public.php#L40

@icewind1991

**Comment by PVince81** [comment]
If we'd leave public links open in maintenance mode, a lot could go wrong. It calls setupFS and all and some public links allow uploading...

**Comment by DeepDiver1975** [comment]
that check is wrong on public.php - there is no user logged in

```php
	public static function checkSingleUserMode() {
		$user = OC_User::getUserSession()->getUser();
		$group = OC_Group::getManager()->get('admin');
		if ($user && \OC::$server->getSystemConfig()->getValue('singleuser', false) && !$group->inGroup($user)) {
			// send http status 503
			header('HTTP/1.1 503 Service Temporarily Unavailable');
			header('Status: 503 Service Temporarily Unavailable');
			header('Retry-After: 120');

			// render error page
			$tmpl = new OC_Template('', 'singleuser.user', 'guest');
			$tmpl->printPage();
			die();
		}
	}

**Comment by DeepDiver1975** [comment]
I'll fix that one ...

**Comment by DeepDiver1975** [comment]
@PVince81 care to retest public.php - THX

**Comment by PVince81** [comment]
So if I understand well, public.php is blocked in the general case, even in single user mode, unless an admin is currently logged in.

Tested:
- [x] public.php not blocked if logged in user is an admin
- [x] public.php is blocked if no user logged in

I found another bug:
- [ ] BUG: log out BEFORE enabling singleuser mode: login page should still work
   1. Logout
   2. Enable single user mode
   3. Clear cache/cookies
   4. => instead of showing the login page it shows the page "single user mode only for admins" with a non-working "Logout" button

Basically this means single user mode only works if the admin was already logged in before switching that mode on. I don't think it's on purpose ?

**Comment by PVince81** [comment]
Just checked master and the bug isn't there: the login page can be displayed in single user mode.
So this is a regression introduced by this PR.

**Comment by DeepDiver1975** [comment]
@PVince81 another try? THX

**Comment by PVince81** [comment]
I'm able to login again, but now even when logged out, public.php isn't blocked any more.
- [ ] BUG: must block public.php when no admin session

**Comment by DeepDiver1975** [comment]
```sh
$ curl http://localhost:8080/public.php
<!DOCTYPE html>
<!--[if lte IE 8]><html class="ng-csp ie ie8 lte9 lte8" data-placeholder-focus="false" lang="en" ><![endif]-->
<!--[if IE 9]><html class="ng-csp ie ie9 lte9" data-placeholder-focus="false" lang="en" ><![endif]-->
<!--[if (gt IE 9)|!(IE)]><!--><html class="ng-csp" data-placeholder-focus="false" lang="en" ><!--<![endif]-->
	<head data-requesttoken="3SoeHaDytw3xayDDOgk2YeRf5NiKbj">
		<meta charset="utf-8">
		<title>
		ownCloud		</title>
		<meta http-equiv="X-UA-Compatible" content="IE=edge">
		<meta name="viewport" content="width=device-width, minimum-scale=1.0, maximum-scale=1.0">
		<meta name="apple-itunes-app" content="app-id=543672169">
		<link rel="shortcut icon" type="image/png" href="/core/img/favicon.png">
		<link rel="apple-touch-icon-precomposed" href="/core/img/favicon-touch.png">
					<link rel="stylesheet" href="/core/css/styles.css?v=98bad1b482bcb42a1d5c916c34645aa2" media="screen">
					<link rel="stylesheet" href="/core/css/header.css?v=98bad1b482bcb42a1d5c916c34645aa2" media="screen">
					<link rel="stylesheet" href="/core/css/mobile.css?v=98bad1b482bcb42a1d5c916c34645aa2" media="screen">
					<link rel="stylesheet" href="/core/css/icons.css?v=98bad1b482bcb42a1d5c916c34645aa2" media="screen">
					<link rel="stylesheet" href="/core/css/fonts.css?v=98bad1b482bcb42a1d5c916c34645aa2" media="screen">
					<link rel="stylesheet" href="/core/css/apps.css?v=98bad1b482bcb42a1d5c916c34645aa2" media="screen">
					<link rel="stylesheet" href="/core/css/fixes.css?v=98bad1b482bcb42a1d5c916c34645aa2" media="screen">
					<link rel="stylesheet" href="/core/css/multiselect.css?v=98bad1b482bcb42a1d5c916c34645aa2" media="screen">
					<link rel="stylesheet" href="/core/vendor/jquery-ui/themes/base/jquery-ui.css?v=98bad1b482bcb42a1d5c916c34645aa2" media="screen">
					<link rel="stylesheet" href="/core/css/jquery-ui-fixes.css?v=98bad1b482bcb42a1d5c916c34645aa2" media="screen">
					<link rel="stylesheet" href="/core/css/jquery-tipsy.css?v=98bad1b482bcb42a1d5c916c34645aa2" media="screen">
					<link rel="stylesheet" href="/core/css/jquery.ocdialog.css?v=98bad1b482bcb42a1d5c916c34645aa2" media="screen">
							<script src="/index.php/core/js/oc.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/vendor/jquery/jquery.min.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/vendor/jquery/jquery-migrate.min.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/vendor/jquery-ui/ui/jquery-ui.custom.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/vendor/underscore/underscore.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/vendor/moment/min/moment-with-locales.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/vendor/handlebars/handlebars.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/vendor/blueimp-md5/js/md5.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/js/placeholders.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/js/jquery-tipsy.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/js/compatibility.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/js/jquery.ocdialog.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/js/oc-dialogs.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/js/js.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/js/l10n.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/js/octemplate.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/js/eventsource.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/js/config.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/search/js/search.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/js/oc-requesttoken.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/js/apps.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/vendor/snapjs/dist/latest/snap.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/js/placeholder.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/js/jquery.avatar.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/js/avatar.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					<script src="/core/js/backgroundjobs.js?v=98bad1b482bcb42a1d5c916c34645aa2"></script>
					</head>
	<body id="body-login">
		<noscript>
	<div id="nojavascript">
		<div>
			This application requires JavaScript for correct operation. Please <a href="http://enable-javascript.com/" target="_blank" rel="noreferrer">enable JavaScript</a> and reload the page.		</div>
	</div>
</noscript>
		<div class="wrapper">
			<div class="v-align">
									<header role="banner">
						<div id="header">
							<div class="logo svg">
								<h1 class="hidden-visually">
									ownCloud								</h1>
							</div>
							<div id="logo-claim" style="display:none;"></div>
						</div>
					</header>
								<ul>
	<li class='update'>
		This ownCloud instance is currently in single user mode.<br><br>
		This means only administrators can use the instance.<br><br>
		Contact your system administrator if this message persists or appeared unexpectedly.		<br><br>
		Thank you for your patience.<br><br>
		<a class="button" href="/index.php?logout=true&requesttoken=3SoeHaDytw3xayDDOgk2YeRf5NiKbj">Log out</a>
	</li>
</ul>
				<div class="push"></div><!-- for sticky footer -->
			</div>
		</div>
		<footer role="contentinfo">
			<p class="info">
				<a href="https://owncloud.org" target="_blank" rel="noreferrer">ownCloud</a> – web services under your control			</p>
		</footer>
	</body>
</html>

**Comment by PVince81** [comment]
Ah, misunderstanding. Try creating a public link (which at some point used to use public.php until we introduced shortened links). The public link can still be opened even when not logged in in single user mode. I think the public link must be blocked, unless there's an admin session open.

**Comment by DeepDiver1975** [comment]
> which at some point used to use public.php until we introduced shortened links

brilliant - it uses index.php meanwhile - fixing it .....

**Comment by DeepDiver1975** [comment]
@PVince81 I see no way how to do this - because the login page would be blocked as well

**Comment by PVince81** [comment]
Ok. I'll say let's move forward and merge what we already have here :+1:

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by MorrisJobke** [comment]
@karlitschek Backport?

#15498 was scheduled for 8.0.4

**Comment by karlitschek** [comment]
please backport
